### ISSUE TITLE
(PRIMA) Arraysection inside Array Reference throws error 

### ISSUE THREAD

**Issue by jinangshah21** [lfortran/lfortran#5516]
MRE : 
```fortran
program test 
integer :: A(3) = [1,2,3]
integer :: X(2) = [1,2]
integer :: Y = 2
A(X(:)) = Y
end program
```

LFortran Error : 
```console
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  Binary file "/mnt/c/lfortran/src/bin/lfortran", in _start
  File "/build/glibc-LcI20x/glibc-2.31/csu/../csu/libc-start.c", line 308, in __libc_start_main
  File "/mnt/c/lfortran/src/bin/lfortran.cpp", line 2725, in main
    return main_app(argc, argv);
  File "/mnt/c/lfortran/src/bin/lfortran.cpp", line 2664, in main_app(int, char**)
    err = compile_src_to_object_file(arg_file, tmp_o, false,
  File "/mnt/c/lfortran/src/bin/lfortran.cpp", line 1070, in (anonymous namespace)::compile_src_to_object_file(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, LCompilers::CompilerOptions&, LCompilers::PassManager&)
    res = fe.get_llvm3(*asr, lpm, diagnostics, infile);
  File "/mnt/c/lfortran/src/lfortran/fortran_evaluator.cpp", line 382, in LCompilers::FortranEvaluator::get_llvm3(LCompilers::ASR::TranslationUnit_t&, LCompilers::PassManager&, LCompilers::diag::Diagnostics&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    compiler_options, run_fn, infile);
  File "/mnt/c/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 10383, in LCompilers::asr_to_llvm(LCompilers::ASR::TranslationUnit_t&, LCompilers::diag::Diagnostics&, llvm::LLVMContext&, Allocator&, LCompilers::PassManager&, LCompilers::CompilerOptions&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    v.visit_asr((ASR::asr_t&)asr);
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 5186, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_asr(LCompilers::ASR::asr_t const&)
    void visit_asr(const asr_t &b) { visit_asr_t(b, self()); }
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 5162, in void LCompilers::ASR::visit_asr_t<LCompilers::ASRToLLVMVisitor>(LCompilers::ASR::asr_t const&, LCompilers::ASRToLLVMVisitor&)
    case asrType::unit: { v.visit_unit((const unit_t &)x); return; }
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 5187, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_unit(LCompilers::ASR::unit_t const&)
    void visit_unit(const unit_t &b) { visit_unit_t(b, self()); }
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 4892, in void LCompilers::ASR::visit_unit_t<LCompilers::ASRToLLVMVisitor>(LCompilers::ASR::unit_t const&, LCompilers::ASRToLLVMVisitor&)
    case unitType::TranslationUnit: { v.visit_TranslationUnit((const TranslationUnit_t &)x); return; }
  File "/mnt/c/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 988, in LCompilers::ASRToLLVMVisitor::visit_TranslationUnit(LCompilers::ASR::TranslationUnit_t const&)
    visit_symbol(*item.second);
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 5189, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_symbol(LCompilers::ASR::symbol_t const&)
    void visit_symbol(const symbol_t &b) { visit_symbol_t(b, self()); }
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 4900, in void LCompilers::ASR::visit_symbol_t<LCompilers::ASRToLLVMVisitor>(LCompilers::ASR::symbol_t const&, LCompilers::ASRToLLVMVisitor&)
    case symbolType::Program: { v.visit_Program((const Program_t &)x); return; }
  File "/mnt/c/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 3292, in LCompilers::ASRToLLVMVisitor::visit_Program(LCompilers::ASR::Program_t const&)
    this->visit_stmt(*x.m_body[i]);
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 5206, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_stmt(LCompilers::ASR::stmt_t const&)
    void visit_stmt(const stmt_t &b) { visit_stmt_t(b, self()); }
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 4926, in void LCompilers::ASR::visit_stmt_t<LCompilers::ASRToLLVMVisitor>(LCompilers::ASR::stmt_t const&, LCompilers::ASRToLLVMVisitor&)
    case stmtType::Assignment: { v.visit_Assignment((const Assignment_t &)x); return; }
  File "/mnt/c/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 5029, in LCompilers::ASRToLLVMVisitor::visit_Assignment(LCompilers::ASR::Assignment_t const&)
    this->visit_expr(*x.m_target);
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 5254, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_expr(LCompilers::ASR::expr_t const&)
    void visit_expr(const expr_t &b) { visit_expr_t(b, self()); }
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 5043, in void LCompilers::ASR::visit_expr_t<LCompilers::ASRToLLVMVisitor>(LCompilers::ASR::expr_t const&, LCompilers::ASRToLLVMVisitor&)
    case exprType::ArrayItem: { v.visit_ArrayItem((const ArrayItem_t &)x); return; }
  File "/mnt/c/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 2365, in LCompilers::ASRToLLVMVisitor::visit_ArrayItem(LCompilers::ASR::ArrayItem_t const&)
    this->visit_expr_wrapper(curr_idx.m_right, true);
  File "/mnt/c/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 5550, in LCompilers::ASRToLLVMVisitor::visit_expr_wrapper(LCompilers::ASR::expr_t*, bool)
    this->visit_expr(*x);
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 5254, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_expr(LCompilers::ASR::expr_t const&)
    void visit_expr(const expr_t &b) { visit_expr_t(b, self()); }
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 5044, in void LCompilers::ASR::visit_expr_t<LCompilers::ASRToLLVMVisitor>(LCompilers::ASR::expr_t const&, LCompilers::ASRToLLVMVisitor&)
    case exprType::ArraySection: { v.visit_ArraySection((const ArraySection_t &)x); return; }
AssertFailed: ASR::is_a<ASR::String_t>(*ASRUtils::expr_type(x.m_v)) && n_dims == 0
```

**Comment by jinangshah21** [comment]
It is similar as #5484 . But here `ArrayReference` is on LHS (Assigned to a value) .

**Comment by AnshMehta1** [comment]
Will work on it.

**Comment by jinangshah21** [comment]
> Will work on it.

I think #5154 is towards this issue. You can refer it.

**Comment by AnshMehta1** [comment]
> > Will work on it.
> 
> I think #5154 is towards this issue. You can refer it.



> > Will work on it.
> 
> I think #5154 is towards this issue. You can refer it.

I understood the issue , will work on it once the PR #5154 is merged into main as this would be an enhancement to that PR.

**Comment by AnshMehta1** [comment]
With with recent main , MRE 
```
program test 
integer :: A(3) = [1,2,3]
integer :: X(2) = [1,2]
integer :: Y = 2
A(X(:)) = Y
print *, A
end program
```

LFortran result : 
```
(lf) lordansh@LordAnsh:~/dev/lfort1/lfortran$ ./src/bin/lfortran try.f90
1    2    3
```

GFortran result 
```
(lf) lordansh@LordAnsh:~/dev/lfort1/lfortran$ gfortran try.f90 && ./a.out
           2           2           3
```

This is to be addressed and fixed . Will work on this.

**Comment by AnshMehta1** [comment]
```diff
diff --git a/src/libasr/pass/array_op.cpp b/src/libasr/pass/array_op.cpp
index bcb9dcca2..afac8c618 100644
--- a/src/libasr/pass/array_op.cpp
+++ b/src/libasr/pass/array_op.cpp
@@ -2189,6 +2189,10 @@ class ArrayOpVisitor : public ASR::CallReplacerOnExpressionsVisitor<ArrayOpVisit
                     result_inc.push_back(al, m_increment);
                 }
                 use_custom_loop_params = true;
+            } else if( ASR::is_a<ASR::ArrayItem_t>(*x.m_target) && ASR::is_a<ASR::ArrayBroadcast_t>(*x.m_value) ) {
+                ASR::ArrayItem_t* array_item = ASR::down_cast<ASR::ArrayItem_t>(x.m_target);
+                replacer.replace_ArrayItem(array_item);
+                replacer.result_type = ASRUtils::expr_type(replacer.result_var);
             }
             remove_original_statement = true;
```

Made these changes to set result_var but still getting some abrupt segfaults.

**Comment by gxyd** [comment]
I applied the diff mentioned here: https://github.com/lfortran/lfortran/issues/5516#issuecomment-2517128754, and ran the program mentioned here: https://github.com/lfortran/lfortran/issues/5516#issuecomment-2514629129.

I don't get any segfault, but the same output as mentioned in the above comment.

Though, usually segfault might indicate an access of value which is `nullptr` (maybe `x.m_target` or `x.m_value`)

**Comment by AnshMehta1** [comment]
Will have a look into it again.

**Comment by gxyd** [comment]
Don't want to bombard you with things, but something to note for maybe later (in current *main*) the MRE mentioned here: https://github.com/lfortran/lfortran/issues/5516#issuecomment-2514629129, when ran as below
```console
> lfortran test.f90
1    2    3
> lfortran test.f90 --experimental-simplifier
ASR verify pass error: ArrayItem::m_type cannot be array.
  --> test.f90:15:5
   |
15 |     A(X(:)) = Y
   |     ^^^^^^^ failed here


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

Probably something to look at as well (maybe as a part of this issue, but later)

**Comment by certik** [comment]
Create a do loop for the LHS as well. Currently we only create a do loop for the RHS for indexing with arrays.

**Comment by Pranavchiku** [comment]
Way forward: https://github.com/lfortran/lfortran/pull/5633/files#r1878536299
